Add support XSPI for S32K5

haduongquang opened the pull request:
Add support for NXP S32 XSPI controller.

Add a shim driver for xspi hyperRAM memory devices on a NXP S32 XSPI bus. 

Enable samples and tests for the S32K5XXCVB on-board IS66WVH8M8DALL memory.

haduongquang committed:
drivers: memc: add NXP S32 XSPI controller for K5

add NXP S32 XSPI controller for K5

Signed-off-by: Ha Duong Quang <ha.duongquang@nxp.com>

haduongquang committed:
drivers: flash: add NXP S32 XSPI flash hyperRAM driver for K5

add NXP S32 XSPI flash hyperRAM driver for K5

Signed-off-by: Ha Duong Quang <ha.duongquang@nxp.com>

haduongquang committed:
samples: enable flash samples for s32k566 board

Enable flash samples for s32k566 board

Signed-off-by: Ha Duong Quang <ha.duongquang@nxp.com>

haduongquang committed:
tests: enable flash tests for s32k566 board

Enable flash tests for s32k566 board

Signed-off-by: Ha Duong Quang <ha.duongquang@nxp.com>

JarmouniA commented on boards/nxp/s32k5xxcvb/s32k5xxcvb_s32k566.dtsi:
```diff
@@ -74,3 +75,84 @@
+		};
+	};
+
+	is66wvh8m8dall: is66wvh8m8dall@0 {
```
```suggestion
	flash_controller: flash-comtroller@0 {
```
About nodes naming: https://devicetree-specification.readthedocs.io/en/latest/chapter2-devicetree-basics.html#generic-names-recommendation

Also, the flash chip should be a child node of the flash controller, see for ex. https://github.com/zephyrproject-rtos/zephyr/blob/e93279e773f95010c495aced866b62baeb0fe1b8/boards/st/stm32h750b_dk/stm32h750b_dk-common.dtsi#L196

manuargue reviewed (changes requested):
@haduongquang please address the twister build failures. Beware that your hal_nxp PR was mistakenly merged prematurely (https://github.com/zephyrproject-rtos/hal_nxp/pull/648#issuecomment-3689792534) so I'm not so sure if the build failures happen because of the manifest may need updates to no longer point to that PR. Worst case, the hal_nxp PR should be reverted back and resume this review with a fresh hal_nxp PR.

haduongquang commented:
> @haduongquang please address the twister build failures. Beware that your hal_nxp PR was mistakenly merged prematurely ([zephyrproject-rtos/hal_nxp#648 (comment)](https://github.com/zephyrproject-rtos/hal_nxp/pull/648#issuecomment-3689792534)) so I'm not so sure if the build failures happen because of the manifest may need updates to no longer point to that PR. Worst case, the hal_nxp PR should be reverted back and resume this review with a fresh hal_nxp PR.

I've fixed it.

haduongquang commented on dts/bindings/mtd/nxp,s32-xspi-device.yaml:
```diff
@@ -0,0 +1,28 @@
+  reg:
+    required: true
+
+  max-program-buffer-size:
```
No, it's different. This properties is the page size for writing (typically 256B-few KB), while erase-block-size is the sector size for erasing (typically much larger, like 8KB-64KB).

haduongquang commented on boards/nxp/s32k5xxcvb/s32k5xxcvb_s32k566.dtsi:
```diff
@@ -74,3 +75,84 @@
+		};
+	};
+
+	is66wvh8m8dall: is66wvh8m8dall@0 {
```
I updated

manuargue commented on drivers/memc/memc_nxp_s32_xspi.h:
```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright 2023 NXP
```
```suggestion
 * Copyright 2025 NXP
```

manuargue commented on dts/bindings/mtd/nxp,s32-xspi-hyperram.yaml:
```diff
@@ -0,0 +1,69 @@
+      The word address of the device ID in ASO (Application-Specific Object).
+      This address specifies the exact location within the memory where the device ID is stored.
+
+  master-locked-type:
```
> Selects the clock types:

```suggestion
  master-clock-type:
```
typo?

haduongquang replied:
right, I updated

JarmouniA commented on boards/nxp/s32k5xxcvb/s32k5xxcvb_s32k566.dtsi:
```diff
@@ -74,3 +75,85 @@
+	};
+
+	/* HyperRAM IS66WVH8M8DAL */
+	hyperram: hyperram@0 {
```
```suggestion
	hyperram: flash-controller@0 {
```

manuargue replied:
```suggestion
	is66wvh8m8dall: flash-controller@0 {
```
I wouldn't remove the node label `is66wvh8m8dall`. @JarmouniA's previous comment was about the node name. And pls remove the comment in line above, the node label is explicit enough.

haduongquang replied:
I updated

haduongquang committed:
boards: s32k5xxcvb: enable flash controller for XSPI

This board has a IS66WVH8M8DALL memory connected to
the only XSPI port available in S32K566.

Signed-off-by: Ha Duong Quang <ha.duongquang@nxp.com>

haduongquang commented on drivers/memc/memc_nxp_s32_xspi.c:
```diff
@@ -0,0 +1,282 @@
+
+#define SFP_FRAD_NODE(n) DT_INST_CHILD(n, sfp_frad)
+
+#define XSPI_ACP_POLICY(node_id)							\
```
I updated

manuargue commented on drivers/memc/memc_nxp_s32_xspi.h:
This is a common pattern with NXP external memory controllers because the hardware IP exposes multiple ports. In Zephyr, we map the main controller as a memc driver and treat each port as a flash controller driver. You’ll see the same approach in other NXP memory controller implementations.

JarmouniA commented on dts/bindings/xspi/nxp,s32-xspi.yaml:
```diff
@@ -0,0 +1,271 @@
+  hold-time-2x:
+    type: boolean
+    description: |
+      Set to align incoming data with 2x serial flash half clock, when in DDR
```
```suggestion
      Align incoming data with 2x serial flash half clock, when in DDR
```

JarmouniA commented on dts/bindings/xspi/nxp,s32-xspi.yaml:
```diff
@@ -0,0 +1,271 @@
+  sample-delay-half-cycle:
+    type: boolean
+    description: |
+      Set to use half-cycle early DQS delay when sampling received data.
```
```suggestion
      Use half-cycle early DQS delay when sampling received data.
```

JarmouniA commented on dts/bindings/xspi/nxp,s32-xspi.yaml:
```diff
@@ -0,0 +1,271 @@
+  ahb-buffers-masters:
+    type: array
+    description: |
+      Masters ID's for the AHB receive buffers. The master ID of every incoming
```
```suggestion
      Master IDs for the AHB receive buffers. The master ID of every incoming
```

JarmouniA commented on dts/bindings/xspi/nxp,s32-xspi.yaml:
```diff
@@ -0,0 +1,271 @@
+      - "BYPASSED"
+      - "AUTO_UPDATE"
+    description: |
+      DLL mode. The supported modes depends on the SoC.
```
```suggestion
      DLL mode. The supported modes depend on the SoC.
```

JarmouniA commented on dts/bindings/xspi/nxp,s32-xspi.yaml:
```diff
@@ -0,0 +1,271 @@
+  lock-configuration:
+    type: boolean
+    description: |
+      Set to lock configuration registers after initialization until next reset.
```
```suggestion
      Lock configuration registers after initialization until next reset.
```

haduongquang commented on dts/bindings/xspi/nxp,s32-xspi.yaml:
```diff
@@ -0,0 +1,271 @@
+  word-addressable:
+    type: boolean
+    description: |
+      True if serial flash is controller is word (2 bytes) addressable.
```
I updated

haduongquang commented on dts/bindings/xspi/nxp,s32-xspi.yaml:
```diff
@@ -0,0 +1,271 @@
+  dword-addressable:
+    type: boolean
+    description: |
+      True if serial flash is controller is word (4 bytes) addressable.
```
I updated

haduongquang commented on dts/bindings/xspi/nxp,s32-xspi.yaml:
```diff
@@ -0,0 +1,271 @@
+  tg1-fixed-priority:
+    type: boolean
+    description: |
+      Set to enable fixed priority for target group 1 (TG1). When disabled,
```
I updated

JarmouniA reviewed (commented):
Nits

Dat-NguyenDuy commented on dts/bindings/xspi/nxp,s32-xspi.yaml:
```diff
@@ -0,0 +1,267 @@
+  a-differential-clock:
+    type: boolean
+    description: |
+      Set to enable differential clock output on CKN pad for side A of the
```
I've updated

Dat-NguyenDuy commented:
> @manuargue please do not resolve my comments.
Hi @JarmouniA, i saw you liked Manuel's replies, so i understood that you agreed with him. I'm not sure if we are on the same page.

Dat-NguyenDuy commented:
Resolved conflicts in the manifest file

Dat-NguyenDuy commented:
Rebased to resolve conflicts

Dat-NguyenDuy commented:
Rebased to resolve conflict

de-nordic commented on drivers/flash/flash_nxp_s32_xspi_hyperram.c:
```diff
@@ -0,0 +1,514 @@
+	size_t erase_size = MIN(FEATURE_XSPI_TX_BUF_SIZE, size);
+	Xspi_Ip_StatusType status;
+
+	if (!area_is_subregion(dev, offset, size) || !size || (offset % erase_size) ||
```
size == 0 is not reason for failure, with size == 0 you are not going to do any work.

Dat-NguyenDuy replied:
I've updated

de-nordic reviewed (changes requested):
https://github.com/zephyrproject-rtos/zephyr/pull/101487#discussion_r2741463720
